### ISSUE TITLE
Registry Substitution Not Substituting Final Variables

### ISSUE THREAD

**Issue by WhiskyTangoFawks** [MinecraftForge/MinecraftForge#3604]
I've been trying to use registry substitution to replace leaves, and what I have discovered is that the substitution is working great, except for one issue, final variables that directly reference the block end up null, rather than substituted.

E.g, in WorldGenTrees, you have 

private static final IBlockState DEFAULT_LEAF = Blocks.LEAVES.getDefaultState().withProperty(BlockOldLeaf.VARIANT, BlockPlanks.EnumType.OAK).withProperty(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
    
This variable ends up null if you try to substitute leaves.

**Comment by LexManos** [comment]
Substitution has nothing to do with that field.
The only part that we would touch is the value held in Blocks.LEAVES.
Its on you to deal with the getDefaultState/withProperty methods correctly to correctly mimic the old blocks.

**Comment by WhiskyTangoFawks** [comment]
I did deal with the get default state method, didn't fix it.  Using the getDefaultState and withProperty methods work perfectly everywhere else.

I did further testing.  I wrote a class with just two variables

	public static final IBlockState stateLeaves = Blocks.LEAVES.getDefaultState();
	public static final Block BlockLeaves = Blocks.LEAVES;

I then initialise an instance of that class **prior** to registering the substitution in pre-init.

In post-init , I run this

		System.out.println("Stored State = " + LeafHolder.stateLeaves.toString());
		System.out.println("Stored Block = " + LeafHolder.BlockLeaves.toString());
		System.out.println("State After Sub= " + Blocks.LEAVES.getDefaultState().toString());
		System.out.println("Block After Sub= " + Blocks.LEAVES.toString());

Which gives me

Stored State = null[check_decay=true,decayable=true,variant=oak]
Stored Block = Block{null}
State After Sub= minecraft:leaves[check_decay=true,decayable=true,variant=oak]
Block After Sub= Block{minecraft:leaves}

And just to confirm, the substitution works fine in game if I get the block in creative and place it in the world.

I'm assuming the same thing is happening in the biome decorator.  It's initialising an instance of the old block prior to the registry substitution happening, which then ends up set to null once the substitution has occurred, (and goes on to generate a bunch of bare trunks in the world).  My own custom coded trees, which get the block from Blocks.LEAVES at the time of them being placed in the world, generate perfectly.
